pkg/k8sgc/k8sgc.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package k8sgc implements a basic for of garbage collection on the Kubernetes API.
+//
+// This job is normally performed by the controller-manager. The component is not available in a test setting, with only
+// the basic API server deployed. This is where k8sgc comes in to fill the gap, so deletion of objects can be tested as
+// expected.
+package k8sgc
+
+import (
+	"context"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type GC interface {
+	Run(ctx context.Context) (bool, error)
+}
+
+type gc struct {
+	client  client.Client
+	toWatch []schema.GroupVersionKind
+}
+
+func New(ctx context.Context, cl client.Client) (GC, error) {
+	var toWatch []schema.GroupVersionKind
+	for gvk := range cl.Scheme().AllKnownTypes() {
+		u := &unstructured.UnstructuredList{}
+		u.SetGroupVersionKind(gvk)
+		err := cl.List(ctx, u)
+		if err != nil {
+			if meta.IsNoMatchError(err) {
+				continue
+			}
+
+			if _, ok := err.(*apierrors.StatusError); ok {
+				continue
+			}
+
+			return nil, err
+		}
+
+		toWatch = append(toWatch, gvk)
+	}
+
+	return &gc{
+		client:  cl,
+		toWatch: toWatch,
+	}, nil
+}
+
+func (g *gc) Run(ctx context.Context) (bool, error) {
```

Comment:
This doesn't collect objects with an ownership loop, but I guess we don't care about that. Not sure it is even permitted.

Reply:
It kind of works, but only if one of the resources in the loop gets deleted first. Strictly speaking, it no longer is a cycle at that point. But I think that is the best we can do.

Reply:
It's completely redundant for our purpose, since we only have a tree for our resources.
```
LinstorCluster   -->   LinstorSatellite   -->   Pods, etc...
             \----->   Deployment, DaemonSet, etc...
```